File: linkedlist/operations_linkedlist.cpp
//https://www.programiz.com/dsa/linked-list-operations
/*
List of basic operations:
1. Traversal
2. Insertion:
    a. at the beginning
    b. at the end
    c. in the middle
3. Deletion:
    a. at the beginning
    b. at the end
    c. in the middle

*/

#include <stdio.h>
#include <iostream> 

using namespace std;

struct Node
{
    /* data */
    int item;
    struct Node* next;
};

//INSERTION
//beginning
void insert_beginning(struct Node** ref, int item){

    //Allocate memory
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));

    //insert item
    new_node->next = (*ref);
    new_node->item = item;

    (*ref) = new_node;
    return;
}

//insert in the middle
void insert(struct Node* prev_node, int item){
    if(prev_node == NULL){
        cout<<"prev node cannot be null"<<endl;
    }
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));

    new_node->next = prev_node->next;
    new_node->item = item;

    prev_node->next = new_node;
    return;

}


//insert at the end
void insert_end(struct Node** ref, int item){
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));

    struct Node* last = (*ref);

    new_node->next = NULL;
    new_node->item = item;
    if (*ref == NULL) {
        *ref = new_node;
        return;
    }

    while (last->next != NULL)
        last = last->next;

    
    last->next = new_node;

    return;


}

void deleteNode(struct Node** ref, int key){
    struct Node *temp = *ref, *prev;
    if (temp != NULL && temp->item == key) {
    *ref = temp->next;
    free(temp);
    return;
    }

    // Find the key to be deleted
    while (temp != NULL && temp->item != key) {
        prev = temp;
        temp = temp->next;

    }
    // If the key is not present
    if (temp == NULL) return;

    // Remove the node
    prev->next = temp->next;

    free(temp);
}

// Print the linked list
void printList(struct Node* node) {
  while (node != NULL) {
    cout << node->item << " ";
    node = node->next;
  }
}


// Driver program
int main() {
  struct Node* head = NULL;

  insert_beginning(&head, 1);
  insert_beginning(&head, 2);
  insert_beginning(&head, 3);
  insert_end(&head, 4);
  insert(head->next, 5);

  cout << "Linked list: ";
  printList(head);

  cout << "\nAfter deleting an element: ";
  deleteNode(&head, 3);
  printList(head);
  cout<<endl;
}

File: linkedlist/operations_linkedlist.py
# https://www.programiz.com/dsa/linked-list-operations
"""
List of basic operations:
1. Traversal
2. Insertion:
    a. at the beginning
    b. at the end
    c. in the middle
3. Deletion:
    a. at the beginning
    b. at the end
    c. in the middle
"""

#Create a node
class Node():
    def __init__(self, item):
        self.item = item
        self.next = None

class LinkedList():
    def __init__(self):
        self.head = None
    
    ###Insertion
    
    #at the beginning
    def insert_beginning(self,item):
        new_node = Node(item)
        new_node.next = self.head
        self.head = new_node
    
    #after a node
    def insert_middle(self, node , data):

        if node is None:
            print("error")
            return
        
        new_node = Node(data)
        new_node.next = node.next
        node.next = new_node
    
    #in the end
    def insert_end(self, item):
        new_node = Node(item)

        if self.head == None:
            self.head = new_node
        
        last = self.head
        while(last.next!=None):
            last = last.next
        last.next = new_node

    ###Deletion
    def deletion(self, position):
        #if empty
        if self.head == None:
            return
        
        temp_node = self.head

        #beginning
        if position == 0:
            temp_node = self.head.next
            temp_node = None
            return
        
        for i in range(position-1):
            temp_node = temp_node.next
            if temp_node is None:
                break
        
        # If the key is not present
        if temp_node is None:
            return
        
        if temp_node.next is None:
            return

        next = temp_node.next.next
        temp_node.next = None
        temp_node.next = next

    def printList(self):
        temp_node = self.head
        while (temp_node):
            print(str(temp_node.item) + " ")
            temp_node = temp_node.next

if __name__ == '__main__':

    list = LinkedList()
    list.insert_beginning(1)
    list.insert_beginning(2)
    list.insert_beginning(3)
    list.insert_end(4)
    list.insert_middle(list.head.next, 5)

    print('Linked list:')
    list.printList()

    print("\nAfter deleting an element:")
    list.deletion(3)
    list.printList()